src/main/java/org/wso2/carbon/inbound/smpp/SMPPListeningConsumer.java

```diff
@@ -0,0 +1,360 @@
+                    + addressRange + ", EnquireLinkTimer : " + enquireLinkTimer
+                    + ", TransactionTimer : " + transactionTimer
+                    + ", ReconnectInterval : " + reconnectInterval
+                    + ", RetryCount : " + retryCount + "for " + name);
+        }
+        logger.info("Initialized the SMPP inbound consumer " + name);
+    }
+
+    /**
+     * Create connection with SMSC and listen to retrieve the messages. Then inject
+     * according to the registered handler.
+     */
+    public void listen() {
+        if (logger.isDebugEnabled()) {
+            logger.debug("Started to Listen SMPP messages for " + name);
+        }
+        try {
+            session = getSession();
+        } catch (IOException e) {
+            reconnectAfter(reconnectInterval);
+            throw new SynapseException("Error while getting the SMPP session", e);
+        }
+    }
+
+    /**
+     * Get the session. If the session still null or not in bound state, then IO
+     * exception will be thrown.
+     *
+     * @return the valid session.
+     * @throws IOException if there is no valid session or session creation is
+     *                     invalid.
+     */
+    private SMPPSession getSession() throws IOException {
+        if (session == null) {
+            logger.info("Initiate session for the first time to " + host + ":" + port + " for " + name);
+            session = newSession();
+        } else if (!session.getSessionState().isBound()) {
+            throw new SynapseException("There is no valid session yet");
+        }
+        return session;
+    }
+
+    /**
+     * Create new SMPPSession.
+     *
+     * @return the {@link SMPPSession}.
+     * @throws IOException if the creation of new session failed.
+     */
+    private SMPPSession newSession() throws IOException {
+        logger.info("Creating new session " + name);
+        bindParameter = new BindParameter(BindType.valueOf(bindType), systemId, password, systemType,
+                TypeOfNumber.valueOf(addressTon), NumberingPlanIndicator.valueOf(addressNpi), addressRange);
+        SMPPSession session = new SMPPSession(host, port, bindParameter);
+        session.setEnquireLinkTimer(enquireLinkTimer);
+        session.setTransactionTimer(transactionTimer);
+        // Set listener to receive notification if there is any session state changes.
+        session.addSessionStateListener(new SessionStateListenerImpl());
+        if (logger.isDebugEnabled()) {
+            logger.debug("Listening SMPP messages for " + name);
+        }
+        // Set listener to receive SMPP messages.
+        session.setMessageReceiverListener(new MessageReceiverListenerImpl());
+        return session;
+    }
+
+    /**
+     * Reconnect session after specified interval.
+     *
+     * @param timeInMillis is the interval.
+     */
+    private void reconnectAfter(final long timeInMillis) {
```

Comment:
Suggestion, can we do something like this instead of current thread handling? https://dzone.com/articles/asynchronous-retry-pattern